driver/src/main/scala/api/MongoConnection.scala

```diff
@@ -412,20 +412,31 @@ object MongoConnection {
 
       // ---
 
+      val (unsupportedKeys, options) = opts
       if (useful.indexOf("@") == -1) {
         val (db, hosts) = parseHostsAndDbName(useful)
-        val (unsupportedKeys, options) = opts
 
-        ParsedURI(hosts, options, unsupportedKeys, db, None)
+        options.authMode match {
+          case X509Authentication => ParsedURI(hosts, options, unsupportedKeys, db,
+            Some(Authenticate(db.getOrElse(throw new URIParsingException(s"Could not parse URI '$uri': authentication information found but no database name in URI")), "", "")))
+          case _ => ParsedURI(hosts, options, unsupportedKeys, db, None)
+        }
       } else {
-        val WithAuth = """([^:]+):([^@]*)@(.+)""".r
+
+        val WithAuth = """([^:]+)(|:[^@]*)@(.+)""".r
+        object PassExtractor {
```

Comment:
I don't think that's required: there will always be a matching group (as it's always `Some`), no extraction, so `drop(1)` is simpler and makes more sense.

Reply:
Regexes are by their nature not very readable and it's not immediately clear that password may start with a colon if you just use `pass.drop(1)`.  Also `pass` is used twice which means that a variable should be used rather than a duplication of `pass.drop(1)`. And I find `stripPrefix(":")` much better / more explicit than `drop(1)`. Having said that if you strongly prefer `drop(1)` please suggest variable names in the pattern match and later and I will refactor.

Reply:
1. Using a custom extract when there is only one `Some` case doesn't make sense for me. Whatever is the normalization operation (`stripPrefix` or  `.drop(1)`), it should be done in the body of the `case WithAuth`.
2. In anyway, the regexp guarantee that either this part is already empty (nothing can be dropped), or it starts with `:`, having the `stripPrefix` checking it again is not useful.

Reply:
stripPrefix doesn't check, it strips the prefix, and provides ready to use pass with the colon removed

I don't want to argue over this really. I can do as below if you wish:

```

        val WithAuth = """([^:]+)(|:[^@]*)@(.+)""".r

        useful match {
          case WithAuth(user, pass, hostsPortsAndDbName) => {
            val (db, hosts) = parseHostsAndDbName(hostsPortsAndDbName)

            db.fold[ParsedURI](throw new URIParsingException(s"Could not parse URI '$uri': authentication information found but no database name in URI")) { database =>

              val passWithoutColon = pass.stripPrefix(":")
              
              if (options.authMode == X509Authentication && passWithoutColon.nonEmpty) {
                throw new URIParsingException("You should not provide a password when authenticating with X509 authentication")
              }

              ParsedURI(hosts, options, unsupportedKeys, Some(database),
                Some(Authenticate(options.authenticationDatabase.
                  getOrElse(database), user, passWithoutColon)))
            }
          }
```

Reply:
https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/immutable/StringLike.scala#L158